### ISSUE TITLE
DEFAULT_HOST not being recognized when using SSL

### ISSUE THREAD

**Issue by patsissons** [nginx-proxy/nginx-proxy#424]
I'm not sure what the exact situation is but I can describe the configuration I'm using and maybe it will trigger some thoughts.

The config works great, except when trying to visit `docker.myhost.com` where it re-routes to the wrong virtual host. I'm not sure why it does that. Both hosts on their own are working great.

obviously `myhost.com` is just a sample hostname, and these are not the full compose configurations, just what i think is relevant.

``` yml
proxy:
  image: 'jwilder/nginx-proxy:latest'
  ports:
    - '80:80'
    - '443:443'
  environment:
    - DEFAULT_HOST=cadvisor.docker.myhost.com

cadvisor:
  expose:
    - '8080'
  environment:
    - VIRTUAL_HOST=cadvisor.docker.myhost.com
    - VIRTUAL_PORT=8080
    - VIRTUAL_PROTO=http

sslhost:
  expose:
    - '443'
  environment:
    - VIRTUAL_HOST=sslhost.docker.myhost.com
    - VIRTUAL_PORT=443
    - VIRTUAL_PROTO=https
    - CERT_NAME=shared
```

**Comment by wader** [comment]
Can you provide some more information like generated nginx config, nginx logs and do you visit using http or https and does it make any difference? does any of the other services not included in the provided config expose any ports or have nginx-proxy related env vars?

**Comment by patsissons** [comment]
going to `http://docker.myhost.com/` will redirect to `https://docker.myhost.com`.

Below is almost verbatim the configs and log. I have to assume it is something to do with the nginx conf file, but what specifically, and why it's not handling the `DEFAULT_HOST` properly is a bit hard for me to diagnose.

```
proxy:
  image: 'jwilder/nginx-proxy:latest'
  environment:
    DEFAULT_HOST: 'cadvisor.docker.marinels.local'
  hostname: proxy.docker.marinels.local
  ports:
    - '80:80'
    - '443:443'
  restart: on-failure
  volumes:
    - '/var/run/docker.sock:/tmp/docker.sock:ro'
    - '/var/lib/docker/volumes/proxy/conf.d:/etc/nginx/conf.d'
    - '/var/lib/docker/volumes/proxy/certs:/etc/nginx/certs'
    - '/var/lib/docker/volumes/ssl:/etc/nginx/certs/shared'

cadvisor:
  image: 'google/cadvisor:latest'
  environment:
    VIRTUAL_HOST: 'cadvisor.docker.marinels.local'
    VIRTUAL_PORT: 8080
    VIRTUAL_PROTO: 'http'
  expose:
    - '8080'
  hostname: cadvisor.docker.marinels.local
  restart: on-failure
  volumes:
    - '/:/rootfs:ro'
    - '/var/run:/var/run'
    - '/sys:/sys:ro'
    - '/var/lib/docker:/var/lib/docker:ro'

gateone:
  image: 'liftoff/gateone:latest'
  environment:
    CERT_NAME: 'marinels'
    GATEONE_REPO_URL: 'https://github.com/patsissons/GateOne.git'
    VIRTUAL_HOST: 'gateone.docker.marinels.local'
    VIRTUAL_PORT: 8000
    VIRTUAL_PROTO: 'https'
  expose:
    - '8000'
  hostname: gateone.docker.marinels.local
  restart: on-failure
  volumes:
    - '/var/lib/docker/volumes/gateone/logs:/gateone/logs'
    - '/var/lib/docker/volumes/gateone/users:/gateone/users'
    - '/var/lib/docker/volumes/gateone/conf.d:/etc/gateone/conf.d'
    - '/var/lib/docker/volumes/gateone/ssl:/etc/gateone/ssl'
    - '/var/lib/docker/volumes/ssl:/etc/gateone/ssl/shared'

gitlab:
  image: 'gitlab/gitlab-ce:latest'
  environment:
    VIRTUAL_HOST: 'gitlab.docker.marinels.local'
    VIRTUAL_PORT: 80
    VIRTUAL_PROTO: 'http'
  expose:
    - '80'
  hostname: 'gitlab.docker.marinels.local'
  ports:
    - '2222:22'
  restart: on-failure
  volumes:
    - '/var/lib/docker/volumes/gitlab/etc:/etc/gitlab'
    - '/var/lib/docker/volumes/gitlab/logs:/var/log/gitlab'
    - '/var/lib/docker/volumes/gitlab/data:/var/opt/gitlab'
```

`/etc/nginx/conf.d/default.conf`
```
# If we receive X-Forwarded-Proto, pass it through; otherwise, pass along the
# scheme used to connect to this server
map $http_x_forwarded_proto $proxy_x_forwarded_proto {
  default $http_x_forwarded_proto;
  ''      $scheme;
}
# If we receive Upgrade, set Connection to "upgrade"; otherwise, delete any
# Connection header that may have been passed to this server
map $http_upgrade $proxy_connection {
  default upgrade;
  '' close;
}
gzip_types text/plain text/css application/javascript application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
log_format vhost '$host $remote_addr - $remote_user [$time_local] '
                 '"$request" $status $body_bytes_sent '
                 '"$http_referer" "$http_user_agent"';
access_log off;
# HTTP 1.1 support
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
server {
        server_name _; # This is just an invalid value which will never trigger on a real hostname.
        listen 80;
        access_log /var/log/nginx/access.log vhost;
        return 503;
}
upstream cadvisor.docker.myhost.com {
                        # cadvisor-1.cadvisor.3559da30
                        server 172.17.0.9:8080;
}
server {
        server_name cadvisor.docker.myhost.com;
        listen 80 default_server;
        access_log /var/log/nginx/access.log vhost;
        location / {
                proxy_pass http://cadvisor.docker.myhost.com;
        }
}
upstream gateone.docker.myhost.com {
                        # gateone-1.gateone.8ca0e833
                        server 172.17.0.2:8000;
}
server {
        server_name gateone.docker.myhost.com;
        listen 80 ;
        access_log /var/log/nginx/access.log vhost;
        return 301 https://$host$request_uri;
}
server {
        server_name gateone.docker.myhost.com;
        listen 443 ssl http2 ;
        access_log /var/log/nginx/access.log vhost;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA
-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-
GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA;
        ssl_prefer_server_ciphers on;
        ssl_session_timeout 5m;
        ssl_session_cache shared:SSL:50m;
        ssl_certificate /etc/nginx/certs/docker.crt;
        ssl_certificate_key /etc/nginx/certs/docker.key;
        add_header Strict-Transport-Security "max-age=31536000";
        location / {
                proxy_pass https://gateone.docker.myhost.com;
        }
}
upstream gitlab.docker.myhost.com {
                        # gitlab-1.gitlab.1b874a45
                        server 172.17.0.3:80;
}
server {
        server_name gitlab.docker.myhost.com;
        listen 80 ;
        access_log /var/log/nginx/access.log vhost;
        location / {
                proxy_pass http://gitlab.docker.myhost.com;
        }
}
```

```
proxy-1 | 2016-04-21T18:40:30.699174149Z forego     | starting nginx.1 on port 5000
proxy-1 | 2016-04-21T18:40:30.701617452Z forego     | starting dockergen.1 on port 5100
proxy-1 | 2016-04-21T18:40:30.809567477Z dockergen.1 | 2016/04/21 18:40:30 Generated '/etc/nginx/conf.d/default.conf' from 4 containers
proxy-1 | 2016-04-21T18:40:30.809877639Z dockergen.1 | 2016/04/21 18:40:30 Running 'nginx -s reload'
proxy-1 | 2016-04-21T18:40:30.846095382Z dockergen.1 | 2016/04/21 18:40:30 Watching docker events
proxy-1 | 2016-04-21T18:40:30.897328582Z dockergen.1 | 2016/04/21 18:40:30 Contents of /etc/nginx/conf.d/default.conf did not change. Skipping notification 'nginx -s reload'
proxy-1 | 2016-04-21T18:41:56.805785722Z nginx.1    | docker.myhost.com 192.168.1.94 - - [21/Apr/2016:18:41:56 +0000] "GET / HTTP/2.0" 200 1191 "-" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36"
proxy-1 | 2016-04-21T18:41:56.886941991Z nginx.1    | docker.myhost.com 192.168.1.94 - - [21/Apr/2016:18:41:56 +0000] "GET /static/gateone.css HTTP/2.0" 304 199 "https://docker.myhost.com/" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36"
proxy-1 | 2016-04-21T18:41:56.909357397Z nginx.1    | docker.myhost.com 192.168.1.94 - - [21/Apr/2016:18:41:56 +0000] "GET /static/gateone.min.js HTTP/2.0" 304 198 "https://docker.myhost.com/" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36"
proxy-1 | 2016-04-21T18:41:56.963895210Z nginx.1    | docker.myhost.com 192.168.1.94 - - [21/Apr/2016:18:41:56 +0000] "GET /auth?check=True HTTP/2.0" 304 156 "https://docker.myhost.com/" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36"
proxy-1 | 2016-04-21T18:42:05.781467025Z nginx.1    | docker.myhost.com 192.168.1.94 - - [21/Apr/2016:18:42:05 +0000] "GET /terminal/static/fonts/ubuntumono-normal.woff HTTP/2.0" 200 107943 "https://docker.myhost.com/" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36"
proxy-1 | 2016-04-21T18:43:06.557999690Z nginx.1    | docker.myhost.com 192.168.1.94 - - [21/Apr/2016:18:43:06 +0000] "GET /terminal/static/fonts/ubuntumono-bold.woff HTTP/2.0" 200 103948 "https://docker.myhost.com/" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36"
```

**Comment by wader** [comment]
Ok! just to be super clear, you visit `http://docker.myhost.com` it redirects to `https://docker.myhost.com` and there the gateone site gets served?

I would guess this happens because the default host has `VIRTUAL_PROTO=http` so only the port 80 server has `default_server`. So when visiting with https (from the redirect) and there is no matching host so the first ssl server is the default. Could that be it?'
Is `docker.crt` a cert with a cname that will match `docker.myhost.com`?

What would you like to happen? `http://docker.myhost.com` redirect to `http://cadvisor.docker.myhost.com` or serve the cadvisor site on `docker.myhost.com` via non-https?

**Comment by wader** [comment]
Something like #298 ?

**Comment by patsissons** [comment]
Your assumptions are correct. `http://docker.myhost.com` redirects to `https://docker.myhost.com` which serves gateone (as it's the first, and only `https` server). `docker.crt` is a shared wildcard SAN cert that authenticates `*.docker.myhost.com`. The issue is that the proxy is automatically redirecting to `https` when there is a valid server on port 80 at the defined `DEFAULT_HOST`. My expectation would be that `http://docker.myhost.com` would serve `DEFAULT_HOST` on the incoming scheme (`http` in this case). It sounds like #298 might fix the issue.

**Comment by patsissons** [comment]
I should mention (now that I have vetted #298) that when setting up the `HTTPS_METHOD` environment variable appropriately, I no longer run into the issue described (because I am not redirecting as I previously was). Not sure if it qualifies a complete resolution of this issue, but it's worth noting for anyone who lands here that you *CAN* solve this behaviour.

**Comment by DevopsDood** [comment]
I need a 444 or something when a host doesn't exist, all existing hosts I want to properly go ssl to where they should.

**Comment by buchdag** [comment]
Fixed in #298

@DevopsDood that was fixed in #2186
